lib/toolbox/strint.c

```diff
@@ -0,0 +1,121 @@
+#include "strint.h"
+
+#include <string.h>
+
+// Splitting out the actual parser helps reduce code size. The manually
+// monomorphized `strint_to_*`s are just wrappers around this generic
+// implementation.
+/**
+ * @brief Converts a string to a `uint64_t` and an auxillary sign bit, checking
+ * the bounds of the integer.
+ * @param [in] str Input string
+ * @param [out] end Pointer to first character after the number in input string
+ * @param [out] abs_out Absolute part of result
+ * @param [out] sign_out Sign part of result (true=negative, false=positive)
+ * @param [in] base Integer base
+ * @param [in] max_abs_negative Largest permissible absolute part of result if
+ * the sign is negative
+ * @param [in] max_positive Largest permissible absolute part of result if the
+ * sign is positive
+ */
+StrintParseError strint_to_uint64_internal(
+    const char* str,
+    char** end,
+    uint64_t* abs_out,
+    bool* sign_out,
```

Comment:
`val_is_negative` or smth similar would be more intuitive - don't have to look up the meaning in docs when calling